[PWA-1947] Gift options on Order Level (Commerce only)

mikhaelbois opened the pull request:
<!--
Before submitting this pull request, please make sure you have read our Contribution Guidelines and your PR meets our contribution standards:
https://github.com/magento/pwa-studio/blob/master/.github/CONTRIBUTING.md

Please fill out as much information as you can about your PR to help speed up the review process.
If your PR addresses an existing GitHub Issue, please refer to it in the title or Additional Information section to make the connection.

We may ask you for changes in your PR in order to meet the standards set in our Contribution Guidelines. PRs that do not comply with our guidelines may be closed at the maintainers' discretion.

Feel free to remove this section before creating this PR. Thank you for your contribution!
-->

## Description

Integration the Gift Message in Order Level feature in the Cart and in the Checkout.
Note that this is currently a Commerce Only feature as the [setGiftOptionsOnCart](https://devdocs.magento.com/guides/v2.4/graphql/mutations/set-gift-options.html) mutation is Commerce Only at the moment.

This is a follow-up of https://github.com/magento/pwa-studio/pull/3435/.

## Related Issue

<!--- This project only accepts pull requests related to open issues -->
<!--- If suggesting a new feature or change, please discuss it in an issue first -->
<!--- If fixing a bug, there should be an issue describing it with steps to reproduce -->
<!--- Please link to the issue here by replacing ISSUE_NUMBER with your actual issue number. -->
<!--- Using the above wording causes Github to automatically close the issue on merge. -->

Closes https://jira.corp.magento.com/browse/PWA-1947.

## Acceptance

<!-- The people and processes this pull request needs before it is merged. -->
<!-- These fields are not required when opening the pull request, but they -->
<!-- should be populated after code review. -->

### Verification Stakeholders

<!-- People who must verify that this solves the attached issue. -->

## Verification Steps

<!-- During code review and QA we will try to ensure there are no bugs introduced by this change -->
<!-- So, we request that you add a detailed test plan on what needs to be checked before this PR gets merged -->
<!-- Feel free to update this after submitting the PR as you discover new scenarios -->
<!-- As part of review/QA we may also add or update the test plan if necessary-->

#### Test scenario(s) for direct fix/feature

<!-- Examples: -->
<!-- 1. Verify user is able to apply filters on category page -->
<!-- 2. Verify user is able to apply filters on search page -->

#### Test scenario(s) for any existing impacted features/areas

Preconditions
1. Adobe Commerce Store with PWA Storefront
2. Gift Messages on order Level is enabled in Stores > Settings > Configuration > Sales> Gift Options
3. Gift Messages for Order Items is disabled in Stores > Settings > Configuration > Sales> Gift Options
4. Gift Wrapping is disabled in Stores > Settings > Configuration > Sales> Gift Options
5. Gift Receipt is enabled
6. Printed Card is enabled

- Users can add gift messages to the oder level (not on order item level )
- User sees no option for gift receipt when Gift Receipt is disabled in Admin Config
- User sees no option for printed card when Printed Card is disabled in Admin Config
- User sees cart prices change when Printed Card has a price defined in Admin Config
- Codebase no longer contains mocked gift options data
- Cypress tests are running

## Checklist

<!--- Go over all the following points, and make sure you've done anything necessary -->

-   I have added tests to cover my changes, if necessary.
-   I have added translations for new strings, if necessary.
-   I have updated the documentation accordingly, if necessary.

pankhurigoel reviewed (approved):
Looks good overall. Thanks.

pankhurigoel reviewed (changes requested):
Looks good overall. Please fix the coverages on following files. `packages/venia-ui/lib/components/CartPage/PriceAdjustments/GiftOptions/giftOptions.js`, `packages/venia-ui/lib/components/CartPage/PriceAdjustments/giftOptionsSection.js`, `packages/venia-ui/lib/components/CartPage/PriceAdjustments/giftOptionsSection.ee.js`

mikhaelbois commented:
@pankhurigoel Tests added. Coverage looks fine now.

pankhurigoel commented:
> yarn run validate-queries

@mikhaelbois try running the `yarn run validate-queries` command locally and fix the issues.

mikhaelbois commented:
@pankhurigoel Here you go. Sorry about that.

pankhurigoel reviewed (dismissed):
Looks perfect!! Thanks.

JustinConabreeAbsolunet commented:
⚠️ QA Smoke test passed. However looking at the database, there's multiple gift message entries created (some partial entries while the user is typing)
⚠️ Cypress tests pass with the exception of the contact page
`Expected to find element: [data-cy="ContactPage-formContainer"] input[name="name"], but never found it.`

mikhaelbois commented:
@JustinConabreeAbsolunet 
I've updated the way the mutation works.
It will now save to cart only when we click on "Review Order" in the Checkout Page.
The other times, it will only save in the cache instead.
It should reduce the number of entries by a lot.

Also fixed the ContactPage Cypress test.

JustinConabreeAbsolunet commented:
@mikhaelbois Getting closer!
✅ Cypress
⚠️ QA issue fixed but there's a new one. Null value saved in backend even when form is untouched. Could be a good idea to keep a "modified" state for the form an only send null if a non-null value was previously saved

mikhaelbois commented:
@JustinConabreeAbsolunet 
Added a comparaison on previous form data and current form data.
If the data is not the same, the form will be submitted.
This way, when you start fresh and don't use the form, it will not be submitted.

JustinConabreeAbsolunet commented:
@mikhaelbois Progress! Now working correctly in Checkout. However, if you add the gift message details in Cart and then go to Checkout, it doesn't get saved when placing the order

pedrochiossi commented on packages/peregrine/lib/talons/CartPage/PriceAdjustments/GiftOptions/useGiftOptions.js:
```diff
@@ -16,145 +20,360 @@ import DEFAULT_OPERATIONS from './giftOptions.gql';
+            cardMessage: cart?.gift_message?.message || ''
+        }),
+        [
+            cart?.gift_message?.from,
```
Can't we just use cart.gift_message as dependencies here?

mikhaelbois replied:
That's what the linter gave me.
From my understanding, if we only use `cart.gift_message`, it won't update if only one of the object entry is updated.
Because it needs to listen to each entry.

pedrochiossi replied:
I see, ok.

pedrochiossi commented on venia-integration-tests/src/fields/contactPage/index.js:
```diff
@@ -1,5 +1,3 @@
-export const contactFormContainer = '[data-cy="ContactPage-root"]';
```
Why remove this selector?

pedrochiossi reviewed (changes requested):
Nicely done @mikhaelbois, I left some comments and suggestions. I think we'll need a sign-off from core team as well.

mikhaelbois commented:
I have found a race condition issue with the checkboxes.
I will take a look on the December 27-28.

jeremycharron commented:
> I have found a race condition issue with the checkboxes. I will take a look on the December 27-28.
> 
> Code Review and QA can still be done in the meantime.

Racing condition has been fixed.
Code review / QA can resume.

fooman commented:
Would have loved to see an attempt to make PriceAdjustments extensible. Currently the approach seems to be that each order total becomes hardcoded.

mikhaelbois commented:
@fooman 
Thanks for the suggestion! I'll let the team know so we can plan a refactor of this section.

michaelyu0123 reviewed (approved):
Just a small change needed, fix it and it LGTM

mikhaelbois commented on packages/venia-ui/i18n/en_US.json:
```diff
@@ -210,9 +210,20 @@
+    "giftOptions.includeGiftMessage": "Optional Message",
     "giftOptions.includeGiftReceipt": "Include gift receipt",
-    "giftOptions.includePrintedCard": "Include printed card",
+    "giftOptions.includePrintedCard": "Add printed card{price}",
```
@michaelyu0123 
The `price` variable is optional so we don't have an unnecessary space in the text if no price is provided.
The space is added when there's a price.

ericeoeur commented:
This has passed all QA testing. Will be merging momentarily. Thanks!
